Add SPI for getting notification when WebsiteDataStore is removed in network process

szewai opened the pull request:
#### 442e878c2eb585b6df5091e0eeb2cb11d694f000
<pre>
Add SPI for getting notification when WebsiteDataStore is removed in network process
<a href="https://bugs.webkit.org/show_bug.cgi?id=256065">https://bugs.webkit.org/show_bug.cgi?id=256065</a>
rdar://problem/108634838

Reviewed by Chris Dumez.

There is a delay between WebsiteDataStore getting destroyed in UI process and network process stopping accessing files
that belong to the WebsiteDataStore. If UI process wants to operate on these files, it has to wait until network process
stops access, to avoid race or corruption. This patch introduces an SPI for client to know when network process finally
stops accessing the files.

API test: WKWebsiteDataStorePrivate.CompletionHandlerForRemovalFromNetworkProcess

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::destroySession):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
(WebKit::NetworkProcess::destroySession):
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setCompletionHandlerForRemovalFromNetworkProcess:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::removeSession):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::~WebsiteDataStore):
(WebKit::WebsiteDataStore::networkProcessDidTerminate):
(WebKit::WebsiteDataStore::setCompletionHandlerForRemovalFromNetworkProcess):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/263603@main">https://commits.webkit.org/263603@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/87e4b8ea5aae2d33b7862e8a4c028330ab93c4e9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5170 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5304 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5484 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6695 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5244 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5167 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5533 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5281 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5520 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5363 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4635 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6713 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2848 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4633 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/10212 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/4713 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4706 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6326 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5134 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4222 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4613 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1243 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8695 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4978 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.h:
```diff
@@ -167,7 +167,7 @@ class NetworkProcess : public AuxiliaryProcess, private DownloadManager::Client,
     void setSession(PAL::SessionID, std::unique_ptr<NetworkSession>&&);
     NetworkSession* networkSession(PAL::SessionID) const final;
-    void destroySession(PAL::SessionID);
+    void destroySession(PAL::SessionID, CompletionHandler<void()>&&);
```
Seems maybe we'd like to add `= [] { }` to avoid updating some call sites.

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -1442,13 +1444,30 @@ void NetworkProcessProxy::removeSession(WebsiteDataStore& websiteDataStore)
+    if (auto existingCompletionHandler = m_removalCompletionHandlers.take(websiteDataStore.sessionID()))
+        existingCompletionHandler("New completion handler is set"_s);
+
+    m_removalCompletionHandlers.set(websiteDataStore.sessionID(), WTFMove(completionHandler));
```
Should use `add()` since you know the key is not present.

szewai replied:
Removed.

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
```diff
@@ -450,6 +451,7 @@ class NetworkProcessProxy final : public AuxiliaryProcessProxy, private ProcessT
 #endif
 
     WeakHashSet<WebsiteDataStore> m_websiteDataStores;
+    HashMap<PAL::SessionID, CompletionHandler<void(String)>> m_removalCompletionHandlers;
```
This name is too generic given its location. What's being removed? I'd suggest something like `m_sessionRemovalCompletionHandlers`.

szewai replied:
Removed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2426,4 +2426,13 @@ void WebsiteDataStore::setProxyConfigData(const API::Data& data, uuid_t proxyIde
+{
+    auto* networkProcess = networkProcessIfExists();
+    if (!networkProcess)
+        return completionHandler("Network process is not running"_s);
```
Doesn't seem like something we ever want to happen. The client is not in charge on the network process. Odds are we should either launch the network process or we should be able to set the handler without launching the network process. This should not be an operation that can fail.

cdumez replied:
My 2 cents. We can probably store the completion handler on the WebsiteDataStore and have `NetworkProcess::removeSession()` "take" it from the WebsiteDataStore. This would avoid having yet another HashMap on the NetworkProcessProxy.

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -592,13 +592,16 @@ void NetworkProcess::destroySession(PAL::SessionID sessionID)
-        storageManager.close([this, protectedThis = Ref { *this }, storageManager = &storageManager]() {
+        storageManager.close([this, protectedThis = Ref { *this }, storageManager = &storageManager, completionHandler = std::exchange(completionHandler, { })]() mutable {
             m_closingStorageManagers.remove(storageManager);
+            completionHandler();
```
It would be good to have some API testing. I am a little worried the IPC back won't reliably make it since we're doing this right before exiting the process.

szewai replied:
Added API test. 
If network process exits, won't UI process error out the pending message reply?

cdumez replied:
Yes, I think that's right. Even though calling the completionHandler on the network process side may not work because we're about to exit. The UIProcess would call the completion handler on its side if there is a pending sendWithAsyncReply() when the IPC connection gets severed.

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -1442,8 +1442,14 @@ void NetworkProcessProxy::removeSession(WebsiteDataStore& websiteDataStore)
-    if (canSendMessage())
-        send(Messages::NetworkProcess::DestroySession { websiteDataStore.sessionID() }, 0);
+    if (canSendMessage()) {
+        if (auto completionHandler = websiteDataStore.takeCompletionHandlerForRemovalFromNetworkProcess()) {
```
nit: I think this may look a bit better if we avoided the duplication:
```
sendWithAsyncReply(Messages::NetworkProcess::DestroySession { websiteDataStore.sessionID() }, [completionHandler = websiteDataStore.takeCompletionHandlerForRemovalFromNetworkProcess()]() mutable {
    if (completionHandler)
        return completionHandler({ });
});
```

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -179,6 +179,7 @@ WebsiteDataStore::~WebsiteDataStore()
     allDataStores().remove(m_sessionID);
     if (m_networkProcess)
         m_networkProcess->removeSession(*this);
```
Maybe we should be passing `std::exchange(m_completionHandlerForRemovalFromNetworkProcess, {})` as second parameter here? It would avoid having the `take*()` function that is a little awkward.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1634,6 +1635,9 @@ void WebsiteDataStore::networkProcessDidTerminate(NetworkProcessProxy& networkPr
     ASSERT(!m_networkProcess || m_networkProcess == &networkProcess);
     m_networkProcess = nullptr;
+
+    if (auto completionHandler = std::exchange(m_completionHandlerForRemovalFromNetworkProcess, { }))
```
This seems inconvenient from the client standpoint no? It means they have to keep resetting their handler whenever the network process crashes? Seems this could easily be avoided on our side now that we store the completion handler of the data store instead of the NetworkProcessProxy? Any reason we don't?

szewai replied:
The purpose of this SPI is to let client know when they could operate (e.g. delete) on files that are being accessed by the network process, so as long as network process is not accessing the files -- either because network process is not running, or session is removed from network process, or network process gets killed -- the completionHandler will be called.

In the known use case, the client does destroy WebsiteDataStore first, so it's probably fine to only call it when WebsiteDataStore is destroyed. I will change it.

cdumez replied:
My worry is that the network process will likely get relaunched right away if it just crashed. I don't even know if we notify the client that the network process relaunched.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2426,4 +2430,21 @@ void WebsiteDataStore::setProxyConfigData(const API::Data& data, uuid_t proxyIde
+void WebsiteDataStore::setCompletionHandlerForRemovalFromNetworkProcess(CompletionHandler<void(String&&)>&& completionHandler)
+{
+    auto* networkProcess = networkProcessIfExists();
+    if (!networkProcess)
```
Why is this code still here? Should no longer be needed.

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -1438,12 +1438,16 @@ void NetworkProcessProxy::addSession(WebsiteDataStore& store, SendParametersToNe
 
-    if (canSendMessage())
-        send(Messages::NetworkProcess::DestroySession { websiteDataStore.sessionID() }, 0);
+    if (canSendMessage()) {
```
Seems there is a branch we don't call the completion handler?

szewai committed:
Add SPI for getting notification when WebsiteDataStore is removed in network process
https://bugs.webkit.org/show_bug.cgi?id=256065
rdar://problem/108634838

Reviewed by Chris Dumez.

There is a delay between WebsiteDataStore getting destroyed in UI process and network process stopping accessing files
that belong to the WebsiteDataStore. If UI process wants to operate on these files, it has to wait until network process
stops access, to avoid race or corruption. This patch introduces an SPI for client to know when network process finally
stops accessing the files.

API test: WKWebsiteDataStorePrivate.CompletionHandlerForRemovalFromNetworkProcess

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::destroySession):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
(WebKit::NetworkProcess::destroySession):
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setCompletionHandlerForRemovalFromNetworkProcess:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::removeSession):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::~WebsiteDataStore):
(WebKit::WebsiteDataStore::networkProcessDidTerminate):
(WebKit::WebsiteDataStore::setCompletionHandlerForRemovalFromNetworkProcess):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/263603@main
